src/pubsub/ua_pubsub_security.c

```diff
@@ -0,0 +1,167 @@
+#include "ua_pubsub_security.h"
+
+#include "ua_pubsub_manager.h"
+#include "ua_util_internal.h"
+
+static
+UA_StatusCode
+needsDecryption(const UA_Logger *logger,
+                const UA_NetworkMessage *networkMessage,
+                const UA_MessageSecurityMode securityMode,
+                UA_Boolean *doDecrypt) {
+
+    UA_Boolean isEncrypted = networkMessage->securityHeader.networkMessageEncrypted;
+    UA_Boolean requiresEncryption = securityMode > UA_MESSAGESECURITYMODE_SIGN;
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    if(isEncrypted && requiresEncryption) {
+        *doDecrypt = UA_TRUE;
+    } else if(!isEncrypted && !requiresEncryption) {
+        *doDecrypt = UA_FALSE;
+    } else {
+        if(isEncrypted) {
+            UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURITYPOLICY,
+                         "PubSub receive. "
+                         "Message is encrypted but ReaderGroup does not expect encryption");
+            retval = UA_STATUSCODE_BADSECURITYMODEINSUFFICIENT;
+        } else {
+            UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURITYPOLICY,
+                         "PubSub receive. "
+                         "Message is not encrypted but ReaderGroup requires encryption");
+            retval = UA_STATUSCODE_BADSECURITYMODEREJECTED;
+        }
+    }
+    return retval;
+}
+
+static
+UA_StatusCode
+needsValidation(const UA_Logger *logger,
+                    const UA_NetworkMessage *networkMessage,
+                    const UA_MessageSecurityMode securityMode,
+                    UA_Boolean *doValidate) {
+
+    UA_Boolean isSigned = networkMessage->securityHeader.networkMessageSigned;
+    UA_Boolean requiresSignature = securityMode > UA_MESSAGESECURITYMODE_NONE;
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    if(isSigned &&
+       requiresSignature) {
+        *doValidate = UA_TRUE;
+    } else if(!isSigned && !requiresSignature) {
+        *doValidate = UA_FALSE;
+    } else {
+
+        if(isSigned) {
+            UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURITYPOLICY,
+                         "PubSub receive. "
+                         "Message is signed but ReaderGroup does not expect signatures");
+            retval = UA_STATUSCODE_BADSECURITYMODEINSUFFICIENT;
+        } else {
+            UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURITYPOLICY,
+                         "PubSub receive. "
+                         "Message is not signed but ReaderGroup requires signature");
+            retval = UA_STATUSCODE_BADSECURITYMODEREJECTED;
+        }
+    }
+    return retval;
+}
+
+UA_StatusCode
+verifyAndDecrypt(const UA_Logger *logger,
+                     UA_MessageSecurityMode securityMode,
+                 UA_ByteString *buffer, const size_t *currentPosition,
+                 const UA_NetworkMessage *currentNetworkMessage,
+                     void *channelContext,
+                 UA_StatusCode (*setNonce)(void *,const UA_ByteString *),
```

Comment:
Its easier to forward the securitypolicy than 5 separate function pointers that all come from the same securitypolicy.

Reply:
I attempted to create an easily to unit-test function that knows as little as possible from the higher level concepts (especially the configuration structures).